File: Classes/PageViewController.m
// PageViewController.m
//
// Kraftstoff


#import "PageViewController.h"
#import "PageCell.h"
#import "PageCellDescription.h"
#import "AppDelegate.h"


@implementation PageViewController



#pragma mark -
#pragma mark View Resize on Keyboard Events (only when visible)



- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillShow:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillHide:)
                                                 name:UIKeyboardWillHideNotification
                                               object:nil];
}


- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];

    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIKeyboardWillShowNotification
                                                  object:nil];

    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                    name:UIKeyboardWillHideNotification
                                                  object:nil];
}


- (void)keyboardWillShow:(NSNotification*)notification
{
    if (keyboardIsVisible == NO) {

        bottomInsetBeforeKeyboard = self.tableView.contentInset.bottom;
        keyboardIsVisible = YES;
    }

    [UIView animateWithDuration:[[notification userInfo][UIKeyboardAnimationDurationUserInfoKey] doubleValue]
                          delay:0.1
                        options:[[notification userInfo][UIKeyboardAnimationCurveUserInfoKey] integerValue]
                     animations:^{

                         CGRect kRect = [[notification userInfo][UIKeyboardFrameEndUserInfoKey] CGRectValue];

                         UIEdgeInsets insets = self.tableView.contentInset;
                         insets.bottom = kRect.size.height;
                         self.tableView.contentInset = insets;
                     }
                     completion:nil];
}


- (void)keyboardWillHide:(NSNotification*)notification
{
    keyboardIsVisible = NO;
}



#pragma mark -
#pragma mark Dismissing the Keyboard



- (void)dismissKeyboardWithCompletion:(void (^)(void))completion
{
    BOOL scrollToTop = (self.tableView.contentOffset.y > 0.0);
    
    [UIView animateWithDuration:scrollToTop ? 0.25 : 0.15
                     animations: ^{
                         
                         NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
                         
                         if (indexPath)
                             [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
                         
                         if (scrollToTop)
                             [self.tableView scrollToRowAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]
                                                   atScrollPosition:UITableViewScrollPositionTop
                                                           animated:NO];

                         UIEdgeInsets insets = self.tableView.contentInset;
                         insets.bottom = bottomInsetBeforeKeyboard;
                         self.tableView.contentInset = insets;
                     }
                     completion: ^(BOOL finished){
                         
                         [self.view endEditing:YES];
                         completion();
                     }];
}



#pragma mark -
#pragma mark Access to Table Cells



- (PageCellDescription*)cellDescriptionForRow:(NSInteger)rowIndex inSection:(NSInteger)sectionIndex
{
    if (tableSections.count <= sectionIndex)
        return nil;

    NSArray *section = tableSections[sectionIndex];

    if (section.count <= rowIndex)
        return nil;

    return section[rowIndex];
}


- (Class)classForRow:(NSInteger)rowIndex inSection:(NSInteger)sectionIndex
{
    return [[self cellDescriptionForRow:rowIndex inSection:sectionIndex] cellClass];
}


- (id)dataForRow:(NSInteger)rowIndex inSection:(NSInteger)sectionIndex
{
    return [[self cellDescriptionForRow:rowIndex inSection:sectionIndex] cellData];
}


- (void)setData:(id)object forRow:(NSInteger)rowIndex inSection:(NSInteger)sectionIndex
{
    [[self cellDescriptionForRow:rowIndex inSection:sectionIndex] setCellData:object];

    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:rowIndex inSection:sectionIndex];
    PageCell *cell = (PageCell*)[self.tableView cellForRowAtIndexPath:indexPath];

    [cell configureForData:object viewController:self tableView:self.tableView indexPath:indexPath];
}



#pragma mark -
#pragma mark Access to Table Sections



- (void)addSectionAtIndex:(NSInteger)sectionIndex withAnimation:(UITableViewRowAnimation)animation
{
    if (tableSections == nil)
        tableSections = [[NSMutableArray alloc] init];

    if (sectionIndex > [tableSections count])
        sectionIndex = [tableSections count];

    [tableSections insertObject:[[NSMutableArray alloc] init] atIndex:sectionIndex];

    if (animation != UITableViewRowAnimationNone)
        [self.tableView insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:animation];

    [self headerSectionsReordered];
}


- (void)removeSectionAtIndex:(NSInteger)sectionIndex withAnimation:(UITableViewRowAnimation)animation
{
    if (sectionIndex < [tableSections count]) {

        [tableSections removeObjectAtIndex:sectionIndex];
        [self headerSectionsReordered];

        if (animation != UITableViewRowAnimationNone)
            [self.tableView deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:animation];
    }
}


- (void)removeAllSectionsWithAnimation:(UITableViewRowAnimation)animation
{
    NSIndexSet *allSections = [NSIndexSet indexSetWithIndexesInRange:NSMakeRange (0, [tableSections count])];

    [tableSections removeAllObjects];
    [self headerSectionsReordered];

    if (animation != UITableViewRowAnimationNone)
        [self.tableView deleteSections:allSections withRowAnimation:animation];
}



#pragma mark -
#pragma mark Access to Table Rows



- (void)addRowAtIndex:(NSInteger)rowIndex
            inSection:(NSInteger)sectionIndex
            cellClass:(Class)class
             cellData:(id)data
        withAnimation:(UITableViewRowAnimation)animation
{
    // Get valid section index and section
    if (tableSections == nil)
        tableSections = [[NSMutableArray alloc] init];

    if ([tableSections count] == 0)
        [self addSectionAtIndex:0 withAnimation:animation];

    if (sectionIndex > [tableSections count] - 1)
        sectionIndex = [tableSections count] - 1;

    NSMutableArray *tableSection = tableSections[sectionIndex];

    // Get valid row index
    if (rowIndex > [tableSection count])
        rowIndex = [tableSection count];

    // Store cell description
    PageCellDescription *description = [[PageCellDescription alloc] initWithCellClass:class andData:data];
    [tableSection insertObject:description atIndex:rowIndex];

    if (animation != UITableViewRowAnimationNone) {

        // If necessary update position for former bottom row of the section
        if (self.tableView.style == UITableViewStyleGrouped)
            if (rowIndex == [tableSection count] - 1 && rowIndex > 0)
                [self setData:[self dataForRow:rowIndex-1 inSection:sectionIndex]
                       forRow:rowIndex-1
                    inSection:sectionIndex];

        // Add row to table
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:rowIndex inSection:sectionIndex];
        [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:animation];
    }
}



- (void)removeRowAtIndex:(NSInteger)rowIndex
               inSection:(NSInteger)sectionIndex
           withAnimation:(UITableViewRowAnimation)animation
{
    if (sectionIndex < [tableSections count]) {

        NSMutableArray *tableSection = tableSections[sectionIndex];

        if (rowIndex < [tableSection count]) {

            [tableSection removeObjectAtIndex:rowIndex];

            if (animation != UITableViewRowAnimationNone) {

                NSIndexPath *indexPath = [NSIndexPath indexPathForRow:rowIndex inSection:sectionIndex];
                [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:animation];
            }
        }
    }
}



#pragma mark -
#pragma mark Properties

@synthesize tableView = _tableView;

- (UITableView *)tableView
{
    return _tableView;
}


- (void)setTableView:(UITableView *)newTableView
{
    _tableView = newTableView;

    [_tableView setDelegate:self];
    [_tableView setDataSource:self];

    if (!self.nibName && !self.view)
        self.view = newTableView;
}


@synthesize constantRowHeight = _constantRowHeight;

- (void)setConstantRowHeight:(BOOL)newValue
{
    _constantRowHeight = newValue;

    // Force change of delegate to indicate changes
    self.tableView.delegate = nil;
    self.tableView.delegate = self;
}


@synthesize useCustomHeaders = _useCustomHeaders;

- (void)setUseCustomHeaders:(BOOL)newValue
{
    _useCustomHeaders = newValue;

    // Force change of delegate to indicate changes
    self.tableView.delegate = nil;
    self.tableView.delegate = self;
}


- (BOOL)respondsToSelector:(SEL)aSelector
{
    if (aSelector == @selector(tableView:heightForRowAtIndexPath:))
        return !self.constantRowHeight;

    if (aSelector == @selector(tableView:viewForHeaderInSection:) || aSelector == @selector(tableView:heightForHeaderInSection:))
        return _useCustomHeaders;

    return [super respondsToSelector:aSelector];
}



#pragma mark -
#pragma mark UITableViewDelegate



- (void)headerSectionsReordered
{
    headerViews = nil;
}


- (UIView *)tableView:(UITableView *)aTableView viewForHeaderInSection:(NSInteger)section
{
    return nil;
}


- (CGFloat)tableView:(UITableView *)aTableView heightForHeaderInSection:(NSInteger)section
{
    NSString *title = [self tableView:aTableView titleForHeaderInSection:section];

    if ([title length] == 0)
        return 0;

    return [[self tableView:aTableView viewForHeaderInSection:section] bounds].size.height;
}


- (CGFloat)tableView:(UITableView *)aTableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return [[self classForRow:indexPath.row inSection:indexPath.section] rowHeight];
}



#pragma mark -
#pragma mark UITableViewDataSource



- (NSInteger)numberOfSectionsInTableView:(UITableView *)aTableView
{
    return [tableSections count];
}


- (NSInteger)tableView:(UITableView *)aTableView numberOfRowsInSection:(NSInteger)section
{
    if (tableSections == nil)
        return 0;

    return [tableSections[section] count];
}


- (NSString *)tableView:(UITableView *)aTableView titleForHeaderInSection:(NSInteger)section
{
    return nil;
}


- (UITableViewCell*)tableView:(UITableView *)aTableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    PageCellDescription *description = [self cellDescriptionForRow:indexPath.row inSection:indexPath.section];

    PageCell *cell = (PageCell*)[aTableView dequeueReusableCellWithIdentifier:[description.cellClass reuseIdentifier]];

    if (cell == nil)
        cell = [[description.cellClass alloc] init];

    [cell configureForData:description.cellData
            viewController:self
                 tableView:aTableView
                 indexPath:indexPath];

    return cell;
}



#pragma mark -
#pragma mark Memory Management



- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: Classes/PageViewController.swift
//
//  PageViewController.swift
//  kraftstoff
//
//  Created by Ingmar Stein on 03.05.15.
//
//

import UIKit

class PageViewController: UITableViewController {

	var tableSections: [[PageCellDescription]] = []

	override func viewDidLoad() {
		super.viewDidLoad()

		tableView.allowsSelectionDuringEditing = true
		tableView.estimatedRowHeight = 44.0
		tableView.rowHeight = UITableView.automaticDimension
	}

	override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
		super.viewWillTransition(to: size, with: coordinator)

		// this fixes a layout issue: the UITextFieldLabels contained in UITextFields are not correctly resized
		self.tableView.reloadData()
	}

	// MARK: - Dismissing the Keyboard

	func dismissKeyboardWithCompletion(_ completion: @escaping () -> Void) {
		let scrollToTop = self.tableView.contentOffset.y > 0.0

		UIViewPropertyAnimator.runningPropertyAnimator(withDuration: scrollToTop ? 0.25 : 0.15, delay: 0, options: [], animations: {
			if let indexPath = self.tableView.indexPathForSelectedRow {
				self.tableView.deselectRow(at: indexPath, animated: false)
				self.tableView.delegate?.tableView?(self.tableView, didDeselectRowAt: indexPath)
			}

			if scrollToTop {
				self.tableView.scrollToRow(at: IndexPath(row: 0, section: 0),
										   at: .top,
										   animated: false)
			}
		}, completion: { _ in
			self.view.endEditing(true)
			completion()
		})
	}

	// MARK: - Access to Table Cells

	func cellDescriptionForRow(_ rowIndex: Int, inSection sectionIndex: Int) -> PageCellDescription? {
		if tableSections.count <= sectionIndex {
			return nil
		}

		let section = tableSections[sectionIndex]

		if section.count <= rowIndex {
			return nil
		}

		return section[rowIndex]
	}

	func classForRow(_ rowIndex: Int, inSection sectionIndex: Int) -> PageCell.Type? {
		return cellDescriptionForRow(rowIndex, inSection: sectionIndex)?.cellClass
	}

	func dataForRow(_ rowIndex: Int, inSection sectionIndex: Int) -> [String: Any]? {
		return cellDescriptionForRow(rowIndex, inSection: sectionIndex)?.cellData
	}

	func setData(_ object: [String: Any], forRow rowIndex: Int, inSection sectionIndex: Int) {
		cellDescriptionForRow(rowIndex, inSection: sectionIndex)?.cellData = object

		let indexPath = IndexPath(row: rowIndex, section: sectionIndex)
		if let cell = self.tableView.cellForRow(at: indexPath) as? PageCell {
			cell.configureForData(object, viewController: self, tableView: self.tableView, indexPath: indexPath)
		}
	}

	// MARK: - Access to Table Sections

	func addSectionAtIndex(_ idx: Int, withAnimation animation: UITableView.RowAnimation) {
		let sectionIndex: Int
		if idx > tableSections.count {
			sectionIndex = tableSections.count
		} else {
			sectionIndex = idx
		}

		tableSections.insert([PageCellDescription](), at: sectionIndex)

		if animation != .none {
			self.tableView.insertSections(IndexSet(integer: sectionIndex), with: animation)
		}
	}

	func removeSectionAtIndex(_ sectionIndex: Int, withAnimation animation: UITableView.RowAnimation) {
		if sectionIndex < tableSections.count {
			tableSections.remove(at: sectionIndex)

			if animation != .none {
				self.tableView.deleteSections(IndexSet(integer: sectionIndex), with: animation)
			}
		}
	}

	func removeAllSectionsWithAnimation(_ animation: UITableView.RowAnimation) {
		tableSections.removeAll(keepingCapacity: false)

		if animation != .none {
			let allSections = IndexSet(integersIn: 0..<tableSections.count)
			self.tableView.deleteSections(allSections, with: animation)
		}
	}

	// MARK: - Access to Table Rows

	func addRowAtIndex(rowIndex rowIdx: Int, inSection sectionIdx: Int, cellClass: PageCell.Type, cellData: [String: Any], withAnimation animation: UITableView.RowAnimation) {
		// Get valid section index and section
		if tableSections.isEmpty {
			addSectionAtIndex(0, withAnimation: animation)
		}

		let sectionIndex: Int
		if sectionIdx > tableSections.count - 1 {
			sectionIndex = tableSections.count - 1
		} else {
			sectionIndex = sectionIdx
		}

		// Get valid row index
		let rowIndex: Int
		if rowIdx > tableSections[sectionIndex].count {
			rowIndex = tableSections[sectionIndex].count
		} else {
			rowIndex = rowIdx
		}

		// Store cell description
		let description = PageCellDescription(cellClass: cellClass, andData: cellData)
		tableSections[sectionIndex].insert(description, at: rowIndex)

		if animation != .none {
			// If necessary update position for former bottom row of the section
			if self.tableView.style == .grouped {
				if rowIndex == tableSections[sectionIndex].count - 1 && rowIndex > 0 {
					setData(dataForRow(rowIndex-1,
						inSection: sectionIndex)!,
						forRow: rowIndex-1,
						inSection: sectionIndex)
				}
			}

			// Add row to table
			let indexPath = IndexPath(row: rowIndex, section: sectionIndex)
			self.tableView.insertRows(at: [indexPath], with: animation)
		}
	}

	func removeRow(at rowIndex: Int, inSection sectionIndex: Int, withAnimation animation: UITableView.RowAnimation) {
		if sectionIndex < tableSections.count {
			if rowIndex < tableSections[sectionIndex].count {
				tableSections[sectionIndex].remove(at: rowIndex)

				if animation != .none {
					let indexPath = IndexPath(row: rowIndex, section: sectionIndex)
					self.tableView.deleteRows(at: [indexPath], with: animation)
				}
			}
		}
	}

	// MARK: - UITableViewDataSource

	override func numberOfSections(in tableView: UITableView) -> Int {
		return tableSections.count
	}

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return tableSections[section].count
	}

	override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		return nil
	}

	override func tableView(_ tableView: UITableView, shouldIndentWhileEditingRowAt indexPath: IndexPath) -> Bool {
		return false
	}

	override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
		return .none
	}

	override func tableView(_ tableView: UITableView, canFocusRowAt indexPath: IndexPath) -> Bool {
		return false
	}

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let description = cellDescriptionForRow(indexPath.row, inSection: indexPath.section)!

		let cellClass = description.cellClass
		let cell = tableView.dequeueReusableCell(withIdentifier: description.cellClass.reuseIdentifier) as? PageCell ?? cellClass.init()

		cell.configureForData(description.cellData, viewController: self, tableView: tableView, indexPath: indexPath)

		return cell
	}

	override func tableView(_ tableView: UITableView, didEndDisplaying cell: UITableViewCell, forRowAt indexPath: IndexPath) {
		guard let pageCell = cell as? PageCell else { return }

		// https://www.fadel.io/blog/posts/ios-performance-tips-you-probably-didnt-know/
		pageCell.reset()
	}

	// MARK: - Programmatically Selecting Table Rows

	func activateCellAtIndexPath(_ indexPath: IndexPath) {
		let cell = self.tableView.cellForRow(at: indexPath)!
		if cell.canBecomeFirstResponder {
			DispatchQueue.main.async {
				cell.becomeFirstResponder()
				self.tableView.beginUpdates()
				self.tableView.endUpdates()
			}
		}
	}

	func deactivateCellAtIndexPath(_ indexPath: IndexPath) {
		let cell = self.tableView.cellForRow(at: indexPath)!
		if cell.canResignFirstResponder {
			DispatchQueue.main.async {
				cell.resignFirstResponder()
				self.tableView.beginUpdates()
				self.tableView.endUpdates()
			}
		}
	}

	func selectRowAtIndexPath(_ indexPath: IndexPath?) {
		if let path = indexPath {
			self.tableView.selectRow(at: path, animated: false, scrollPosition: .none)
			self.tableView(self.tableView, didSelectRowAt: path)
		}
	}

}
